### ISSUE TITLE
50k-60k infer/sec limitation

### ISSUE THREAD

**Issue by v-hyhyniak-crt** [triton-inference-server/server#7590]
**Description**
Hello NVIDIA Team,

During tests, we have seen a limit of ~50k-60k infer/sec per tritonserver instance regardless of the complexity of the model, number of model instances, and with only partial CPU and GPU utilization.

**Triton Information**
Setup in kubernetes:
* tritonserver: nvcr.io/nvidia/tritonserver:24.06-py3, 20 cpu, 32G mem
* perf-analyzers (x4): nvcr.io/nvidia/tritonserver:24.06-py3-sdk, 20 cpu, 32G mem

**To Reproduce**
* Using any simple model (to avoid heavy inference computation), in my case it was a simple ONNX model that converts int32[1] to float32[1] (with and without batching).
* Model config: [config.txt](https://github.com/user-attachments/files/16851948/config.txt), tested with 1-2-4-8 instance count.
* Starting tritonserver in K8s container: `tritonserver --model-repository $REPO_PATH --model-control-mode=poll --repository-poll-secs 10`
* Testing with 1-4 simultaneous perf-analyzers in K8s containers: `perf_analyzer -m model_name --percentile=80 --request-rate-range=12000:40000:1000 -a -u [tritonserver_ip]:8001 -i grpc --input-data random --measurement-interval 6000 --max-threads=32 --string-length=16 --collect-metrics --metrics-url [tritonserver_ip]:8002/metrics --verbose-csv -v`. Each perf-analyzer used alone was able to generate 25k-30k infer/sec load, so having 4 of them simultaneously should be enough to get loads higher than 60k infer/sec

**Expected behavior**
As long as the resources are partially used (max 5-6 cpu out of 20 available), we would expect an increase of infer/sec when increasing instance count (ic).  In our tests, we had: _1 ic: ~25k infer/sec | 2 ic: ~50k infer/sec | 4 ic: ~50k infer/sec | 8 ic: ~54k infer/sec._. 

We have also tested different setups and options, some of them:
* Adding 2nd similar simple model and querying both of them resulted in the same total ~50-60k infer/sec
* No batching, dynamic batches (1-500) do not help
* Testing different environments:
    * in the docker on the bare machine (without K8s) - _same threshold_
    * in the Marathon application (container orchestration platform for Mesos) - _no such threshold_

:grey_exclamation:  running 2 tritonserver instances in the same container (with `--reuse-grpc-port`) actually doubles the throughput to 100k infer/sec. This may imply that there is no container- or resource-related bottleneck, but rather a limitation inside the tritonserver itself.

:question: Taking above-stated into account, I was wondering if you could please give more information about:
* limits on the large number of inference requests due to Triton itself and not related to inference computation whatsoever
* options/env var to increase the throughput. Maybe something like number of processing threads, grpc handling threads etc
* tools (apart from perf-analyzer or model-analyzer) for profiling/monitoring tritonserver that could help us to find the bottleneck

Thank you!

**Comment by nicolasnoble** [comment]
So we're able to reproduce the caps here, with the amount of QPS you're reporting. We've identified a few areas of improvements where the codebase is indeed bottlenecking and we have action items to explore how we can make this better.

However, I would like to mention a bit that in any network server system, infinitely linear QPS for as long as CPU is available is impossible. There is always a hard cap eventually, without 100% CPU usage, caused generally by lock contentions. The system has to grab various locks to mutate queues, to allocate and deallocate memory, or even to access the network devices. So, looking at CPU/GPU usage as the metric for the upper QPS limit with an empty request is a flawed measure.

Typically, with current hardware and network devices, using TCP and ethernet, the current speed of light limit of a very well thought server system and a very light protocol is around 5MQPS on a 32 cores machine, and CPU will not be maxed out. This still indicates we are at least one order of magnitude slower than where we should be right now.

**Comment by jcuquemelle** [comment]
Can’t wait to have access to a quantum device that would allow us to overcome the speed of light :slightly_smiling_face: 

Joke aside, we are aware that we can’t expect linear scaling of QPS w.r.t Cpu usage, however we could expect to be on the same order of magnitude as what the GRPC stack is able to achieve; According to the official benchmarks here: [Grafana](https://grafana-dot-grpc-testing.appspot.com/?orgId=1), we should expect (for unary requests):
- ~110k QPS on  8 core system (actually using less than 7 cores)
- ~400k QPS on a 32 cores system (actually using 25 cores).

Interestingly enough, the Java setup is almost twice as fast as the C++ one thinking face , and at that throughput, we still have a monotonic (if not linear) relation between QPS and CPU usage.

# TL;DR:
We hit several bottlenecks actually, and more often than not, fixing one issue is not immediately visible because it unlocks another issue to be prevalent, hence the need to regularly watch server-side metrics (and the evolution thereof with increasing load) to properly understand where the next bottleneck is.

We can around 100k QPS on a more complex model than the one mentioned in the original post, by using build-in features:
- additional instance groups instead of more instances in the same group
- Tune the queue delay to maintain bigger batches
- Tune GPRC parameters

We identified a few unintuitive behaviors that are listed at the end of this comment (some of them being often used as first advise for perf improvement, which can hence be very misleading)

We make a few proposals for added monitoring to be able to spot these issues faster

# Full Story
On a similar but more complex experiment than the one of the original post (involving a model with 150+ small input tensors), we identified that we were bottlenecked by the fact that the dynamic batching scheduler is single-threaded, and as such could not build batches fast enough to correctly feed the model instances. In fact, as dynamic batches implies building batches for each tensor individually, having a large number of tensors to batch is very ineffective (we only reached ~5k QPS on that setup). Once we solved this bottleneck by stacking all our input tensors into a single tensor per type, we were back to 50k QPS on that use case.

The question was then, whether we were still hit by this single-threaded dynamic-batching scheduler ? In fact, there was a bit of ambiguity in various sources we found regarding scheduling implementations (1 scheduler per model, one scheduler per model instance, …); in fact the official documentation is quite clear about it: there is 1 scheduler per **instance group**. All examples in the doc mention creating instance groups to target several devices, or assigning priority to models, but in fact it also allows to simply add schedulers to be able to consume grpc request faster by repeating instance groups in the config.pbtxt file: 
```
# this allows to have one queue and one dynamic batcher per model instance
instance_group [{
  count:1
}]
instance_group [{
  count:1
}]

``` 
There was a small caveat here, as doubling the number of model instances resulted in (dynamic) batch sizes to be divided by 2, which induced less efficient Gpu computation, which induced competition between 2 model instances for the Gpu yielding a new bottleneck.
This was alleviated by adding delay on the scheduler queue to be able to have bigger batches:
```
dynamic_batching {
   # this allows to avoid having too small batches which would cause separate model instances to compete for the 
   # GPU because of unefficent gpu usage on small batches (inference time for a batch is ~0.5ms irrespective of 
   # the size of the batch)
   max_queue_delay_microseconds: 500
}
```

With this setting, we reach **92k QPS, P50 latency 2.5ms, P99 latency 7ms, GPU usage 45%, CPU Usage 7 cores**

This is much closer to what the Grpc benchmarks give, and it seems perfectly legit given the fact that here we are actually inferencing on a Gpu. However, due to physical cores / hyperthread ambiguity we though we still had some reserve, and as such we tried putting a third model group, which brought no improvement, even by waiting longer in the queues to have bigger batches.

The next hypothesis  was then that the bottleneck came before the schedulers, hence the GRPC stack. Looking quickly on GRPC performance best practices, we noted 2 parameters we could tweak to check this hypothesis: 
- `GRPC_ARG_MAX_CONCURRENT_STREAMS` (max number of concurrent streams (as we are doing unary calls this is equivalent to requests) handled by the server, default value 100)
- `--grpc-infer-allocation-pool-size` (protobuf message pool size, reusing protobuf messages allows saving memory allocations, default value 8)

Setting both parameters to 1024 and adding a third model group allowed to bump the throughput to: 
**101k QPS, P50 latency 1.9ms, P99 latency 4.6 ms, GPU 45%, CPU 6.8 Vcores**

Quickly sweeping other values on these parameters (100k QPS @ 5ms latency show we would need to handle ~500 concurrent requests) did not allow to go higher. Since we are close to what the official GRPC benchmark say (our server runs on an 16 cores machine, but only uses 8 cores), we decided to stop the investigation (at least on this env / hardware combination).  

## Side note about Numa locality
A bit off topic from the rest of the issues, but we also noticed a very large impact of missed Numa locality: trying to increase the number of Cpus as an experiment to increase throughput, we ended up at some point to have Cpus not located in the same numa node (8 in one node, 2 in the other), this yielded a drop from 90k to ~ 65k QPS. As such, sizing Triton deployments should be made in accordance with the underlying hardware architecture (but fortunately, the triton server provides CLI options to handle this ([Optimization — NVIDIA Triton Inference Server](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/optimization.html#numa-optimization) ), e.g. 
```
tritonserver --host-policy=gpu_0,numa-node=0 --host-policy=gpu_0,cpu-cores=0-15  
```

# Unintuitive behaviors: 
- One might be bottlenecked by the load-testing system. We already saw this with Locust, for which we needed to to distributed load testing but it is not always very practical. Switching to perf_analyser helped, but we still encountered issues with the perf_analyser not being able to keep up with more than 40k QPS because of a memory leak; We were advised to revert to container version 24.02 (before apparition of the leak). Still, reaching 70k+ Qps on the perf_analyser shows limits (at least on our setup), and we had to run 2 separate instances to overcome this limit.
- GRPC client does not actually send requests if GRPC_ARG_MAX_CONCURRENT_STREAMS requests are currently handled by the server
-- This can give the false impression that the client is bottllenecking, in fact it is a back-pressure mechanism to hold requests on the client.
- There is no “global” GRPC queue, so consuming requests is directly done by the model schedulers. The single threaded scheduler issue is then difficult to spot, as no input queue is actually filling.
- Adding model instances can result in a lesser performance if it causes smaller batch sizes, as it makes models instances compete for compute resources because of the reduced efficiency with smaller batches.
- Adding instances in the same model group will not help in cases where batches are not built fast enough for the model to be efficient, you need to add instance groups which come with their own scheduler.
- Adding Cpus to a triton server might decrease performance if Numa topology is not taken into account.

# Suggestions for easier spotting this kind of bottlenecks: 
- have a metric reporting the number of concurrent active streams on GRPC
- have a metric reporting the Cpu usage of each scheduler. Seeing this at 1 full core usage per instance group would directly point to the need for more instance groups
- include this use-case in the Optimization documentation. All pieces are here more or less implicitly, but gathering all of them didn’t occur since opening this ticket.
